File: docs/src/test/scala/docs/stream/operators/sink/WatchTermination.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * license agreements; and to You under the Apache License, version 2.0:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * This file is part of the Apache Pekko project, which was derived from Akka.
 */

package docs.stream.operators.sink

import org.apache.pekko
import pekko.actor.ActorSystem
import pekko.stream.scaladsl.{ Keep, Sink, Source }

import scala.concurrent.ExecutionContext
import scala.util.{ Failure, Success }

object WatchTermination {

  def watchTerminationExample(): Unit = {
    implicit val system: ActorSystem = ???
    implicit val ec: ExecutionContext = ???

    // #watchTermination
    // Wrap Sink.ignore with termination watching
    val terminationWatchedSink =
      Sink.watchTermination(Sink.ignore) { (mat, terminationFuture) =>
        terminationFuture.onComplete {
          case Success(_)         => println("Stream completed successfully")
          case Failure(exception) => println(s"Stream failed: ${exception.getMessage}")
        }
        mat
      }

    Source(1 to 5).runWith(terminationWatchedSink)
    /*
    Prints:
    Stream completed successfully
     */

    // Combine the original materialized value with the termination signal
    val sinkWithBoth =
      Sink.watchTermination(Sink.seq[Int])(Keep.both)

    val (seqFuture, terminationFuture) =
      Source(1 to 3).runWith(sinkWithBoth)

    seqFuture.onComplete {
      case Success(seq) => println(s"Collected: $seq")
      case Failure(ex)  => println(s"Failed to collect: ${ex.getMessage}")
    }

    terminationFuture.onComplete {
      case Success(_)  => println("Stream terminated normally")
      case Failure(ex) => println(s"Stream terminated with error: ${ex.getMessage}")
    }
    /*
    Prints:
    Collected: Vector(1, 2, 3)
    Stream terminated normally
     */
    // #watchTermination
  }
}

File: docs/src/test/java/jdocs/stream/operators/sink/WatchTermination.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * license agreements; and to You under the Apache License, version 2.0:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * This file is part of the Apache Pekko project, which was derived from Akka.
 */

package jdocs.stream.operators.sink;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletionStage;
import org.apache.pekko.Done;
import org.apache.pekko.actor.ActorSystem;
import org.apache.pekko.japi.Pair;
import org.apache.pekko.stream.javadsl.Sink;
import org.apache.pekko.stream.javadsl.Source;

public class WatchTermination {

  static final ActorSystem system = ActorSystem.create("WatchTerminationExample");

  void watchTerminationExample() {
    // #watchTermination
    // Wrap Sink.ignore with termination watching
    Sink<Integer, CompletionStage<Done>> terminationWatchedSink =
        Sink.watchTermination(
            Sink.ignore(),
            (mat, completionStage) -> {
              completionStage.whenComplete(
                  (done, exc) -> {
                    if (exc != null) {
                      System.out.println("Stream failed: " + exc.getMessage());
                    } else {
                      System.out.println("Stream completed successfully");
                    }
                  });
              return completionStage;
            });

    Source.from(Arrays.asList(1, 2, 3, 4, 5)).runWith(terminationWatchedSink, system);
    /*
    Prints:
    Stream completed successfully
     */

    // Combine the original materialized value with the termination signal
    Sink<Integer, Pair<CompletionStage<List<Integer>>, CompletionStage<Done>>> sinkWithBoth =
        Sink.watchTermination(Sink.seq(), Pair::create);

    Pair<CompletionStage<List<Integer>>, CompletionStage<Done>> result =
        Source.from(Arrays.asList(1, 2, 3)).runWith(sinkWithBoth, system);

    result.first().thenAccept(seq -> System.out.println("Collected: " + seq));

    result
        .second()
        .whenComplete(
            (done, exc) -> {
              if (exc != null) {
                System.out.println("Stream terminated with error: " + exc.getMessage());
              } else {
                System.out.println("Stream terminated normally");
              }
            });
    /*
    Prints:
    Collected: [1, 2, 3]
    Stream terminated normally
     */
    // #watchTermination
  }
}
